### DISCUSSION TITLE
Design challenges with Stdin and Signals

### DISCUSSION THREAD

**Discussion by jeff-hykin** [denoland/deno#14197] [Q&A]
"on SIGINT -> do something" can be accomplished with `Deno.addSignalListener("SIGINT", ()=>/*do something*/)`

"on Stdin -> do something" however, has been a bit more challenging. 

The behavior I'm looking for is a callback that isn't put on the event loop. E.g. if the rest of the program finished => Deno.exit (e.g. don't sit and wait for Stdin). 

Is this possible? I've tried a few round-about methods with streams, but haven't found a non-promise based API.

**Comment by jsejcksn** [comment]
The kind of implicit behavior that you describe is not currently possible, and would require the ability to tell Deno to "unref" the stream. The conventional approach is to be explicit — [exit](https://doc.deno.land/deno/stable/~/Deno.exit) your program when it is time to terminate:

```ts
Deno.exit();
```

<details>
<summary>Here's an example of handling <code>stdin</code> "events" with a callback:</summary>

`./stdin-event.ts`:
```ts
/// <reference lib="deno.unstable" />

import { delay } from "https://deno.land/std@0.144.0/async/delay.ts";

async function handleInput(
  callback: (chunk: Uint8Array) => void,
): Promise<void> {
  const isAtty = Deno.isatty(Deno.stdin.rid);
  if (isAtty) Deno.setRaw(Deno.stdin.rid, true, { cbreak: true });
  for await (const chunk of Deno.stdin.readable) callback(chunk);
  if (isAtty) Deno.setRaw(Deno.stdin.rid, false);
}

async function main() {
  console.log("🏁", "Start");

  handleInput((chunk) => console.log("⌨️", { chunk }))
    .catch((ex) => {
      if (ex instanceof Deno.errors.BadResource) {
        console.warn("stdin is closed");
      } else throw ex;
    });

  // wait 5s to simulate some other work
  for (let i = 5; i > 0; i -= 1) {
    console.log("⏱️", i);
    await delay(1e3);
  }

  console.log("🏁", "End");
  Deno.exit(0);
}

main();

```

Then I ran the program and typed `hello` while it was running:

```
% deno run --unstable --check stdin-event.ts
🏁 Start
⏱️ 5
⌨️ { chunk: Uint8Array(1) [ 104 ] }
⌨️ { chunk: Uint8Array(1) [ 101 ] }
⏱️ 4
⌨️ { chunk: Uint8Array(1) [ 108 ] }
⌨️ { chunk: Uint8Array(1) [ 108 ] }
⌨️ { chunk: Uint8Array(1) [ 111 ] }
⏱️ 3
⏱️ 2
⏱️ 1
🏁 End

```

</details>

**Reply by jeff-hykin** [reply]
I see. That is unfortunate as I'm making a package so I won't know what the user of the package intends to do (e.g. if all their settimeouts and promises were resolved or not). 

Thanks for the info
